lsif/src/cache.ts

```diff
@@ -0,0 +1,250 @@
+     * a non-zero `readers` count.
+     *
+     * @param entry The cache entry.
+     * @param value The cache entry's resolved value.
+     */
+    private resolved(entry: CacheEntry<K, V>, value: V): void {
+        const size = this.sizeFunction(value)
+        this.size += size
+        entry.size = size
+
+        let node = this.lruList.tail
+        while (this.size > this.max && node) {
+            const {
+                prev,
+                value: { promise, size, readers },
+            } = node
+
+            if (readers === 0) {
+                this.size -= size
+                this.lruList.removeNode(node)
+                this.cache.delete(node.value.key)
+                promise.then(value => this.disposeFunction(value), () => {})
+            }
+
+            node = prev
+        }
+    }
+}
+
+/**
+ * A cache of SQLite database connections indexed by database filenames.
+ */
+export class ConnectionCache extends GenericCache<string, Connection> {
+    /**
+     * Create a new `ConnectionCache` with the given maximum (soft) size for
+     * all items in the cache.
+     */
+    constructor(max: number) {
+        super(
+            max,
+            // Each handle is roughly the same size.
+            () => 1,
+            // Close the underlying file handle on cache eviction.
+            (connection: Connection) => connection.close()
+        )
+    }
+
+    /**
+     * Invoke `callback` with a SQLite connection object obtained from the
+     * cache or created on cache miss. This connection is guaranteed not to
+     * be disposed by cache eviction while the callback is active.
+     *
+     * @param database The database filename.
+     * @param entities The set of expected entities present in this schema.
+     * @param callback The function invoke with the SQLite connection.
+     */
+    public withConnection<T>(
+        database: string,
+        entities: Function[], // eslint-disable-line @typescript-eslint/ban-types
+        callback: (connection: Connection) => Promise<T>
+    ): Promise<T> {
+        const factory = (): Promise<Connection> =>
+            createConnection({
+                database,
+                entities,
+                type: 'sqlite',
+                name: database,
+                synchronize: true,
+                // logging: 'all',
+            })
+
+        return this.withValue(database, factory, callback)
+    }
+
+    /**
+     * Like `withConnection`, but will open a transaction on the connection
+     * before invoking the callback.
+     *
+     * @param database The database filename.
+     * @param entities The set of expected entities present in this schema.
+     * @param callback The function invoke with a SQLite transaction connection.
+     */
+    public withTransactionalEntityManager<T>(
+        database: string,
+        entities: Function[], // eslint-disable-line @typescript-eslint/ban-types
```

Comment:
```suggestion
        // Decorators are not possible type check
        // eslint-disable-next-line @typescript-eslint/ban-types
        entities: Function[],
```
https://github.com/sourcegraph/eslint-config#when-to-use-eslint-disable

Reply:
Added note.